sources/dma_driver.c

```diff
@@ -0,0 +1,719 @@
+
+// Initialize the DMA stream 
+void dma_stream_init(
+    DMA_TypeDef *dma, 
+    DMA_Stream_TypeDef *dma_stream, 
+    dma_channel_t channel, 
+    dma_direction_t dir, 
+    dma_cm_t cm, 
+    dma_priority_t priority, 
+    dma_addr_inc_mode_t minc, 
+    dma_addr_inc_mode_t pinc, 
+    dma_data_size_t msize, 
+    dma_data_size_t psize)
+{
+    // Enable the DMA clock
+    if (dma == DMA1)
+        RCC->AHB1ENR |= (SET_BIT << SHIFT_21); 
+    else if (dma == DMA2)
+        RCC->AHB1ENR |= (SET_BIT << SHIFT_22); 
+    
+    // Disable the stream 
+    dma_stream_disable(dma_stream); 
+    
+    // Clear all the stream interrupt flags LISR and HISR registers 
+    dma_clear_int_flags(dma); 
+    
+    // Select the DMA channel 
+    dma_chsel(dma_stream, channel); 
+
+    // Configure the direction 
+    dma_dir(dma_stream, dir); 
+
+    // Configure circular 
+    dma_cm(dma_stream, cm); 
+    
+    // Configure the stream priority 
+    dma_priority(dma_stream, priority); 
+    
+    // Configure increment/fixed memory mode 
+    dma_minc(dma_stream, minc); 
+    dma_pinc(dma_stream, pinc); 
+    
+    // Configure data widths 
+    dma_msize(dma_stream, msize); 
+    dma_psize(dma_stream, psize); 
+} 
+
+
+// Configure the DMA stream 
+void dma_stream_config(
+    DMA_Stream_TypeDef *dma_stream, 
+    uint32_t per_addr, 
+    uint32_t mem_addr, 
+    uint16_t data_items)
+{
+    // Configure the total number of data items to be transferred 
+    dma_ndt(dma_stream, data_items); 
+
+    // Set the peripheral port address 
+    dma_par(dma_stream, per_addr); 
+
+    // Set the memory address and subsequently double buffer mode if needed 
+    dma_m0ar(dma_stream, mem_addr); 
+
+    // Enable the DMA stream 
+    dma_stream_enable(dma_stream); 
+}
+
+
+// Configure the FIFO usage 
+void dma_fifo_config(
+    DMA_Stream_TypeDef *dma_stream, 
+    dma_fifo_mode_t mode, 
+    dma_fifo_threshold_t fth, 
+    dma_feie_t feie)
+{
+    // FIFO mode 
+    dma_dmdis(dma_stream, mode); 
+
+    // FIFO threshold 
+    dma_fth(dma_stream, fth); 
+
+    // FIFO interrupt enable 
+    dma_feie(dma_stream, feie); 
+}
+
+
+// Configure the DMA interrupts 
+void dma_int_config(
+    DMA_Stream_TypeDef *dma_stream,  
+    dma_tcie_t tcie, 
+    dma_htie_t htie, 
+    dma_teie_t teie, 
+    dma_dmeie_t dmeie)
+{
+    // Transfer complete interrupt 
+    dma_tcie(dma_stream, tcie); 
+
+    // Half transfer interrupt 
+    dma_htie(dma_stream, htie); 
+
+    // Transfer error interrupt 
+    dma_teie(dma_stream, teie); 
+
+    // Direct mode error interrupt 
+    dma_dmeie(dma_stream, dmeie); 
+}
+
+//================================================================================
+
+
+//================================================================================
+// DMA interrupt status registers 
+
+// Clear all the stream interrupt flags LISR and HISR registers 
+void dma_clear_int_flags(
+    DMA_TypeDef *dma)
+{
+    dma->LIFCR = ~((uint32_t)CLEAR); 
```

Comment:
Now this is something that is a problem with C the language, but you should pretty much _always_ check if the pointer parameter is not NULL before using it in a function.

Eg, 'if(dmg != NULL)'